README.md

[![Arduino CI](https://github.com/RobTillaart/DHT20/workflows/Arduino%20CI/badge.svg)](https://github.com/marketplace/actions/arduino_ci)
[![Arduino-lint](https://github.com/RobTillaart/DHT20/actions/workflows/arduino-lint.yml/badge.svg)](https://github.com/RobTillaart/DHT20/actions/workflows/arduino-lint.yml)
[![JSON check](https://github.com/RobTillaart/DHT20/actions/workflows/jsoncheck.yml/badge.svg)](https://github.com/RobTillaart/DHT20/actions/workflows/jsoncheck.yml)
[![License: MIT](https://img.shields.io/badge/license-MIT-green.svg)](https://github.com/RobTillaart/DHT20/blob/master/LICENSE)
[![GitHub release](https://img.shields.io/github/release/RobTillaart/DHT20.svg?maxAge=3600)](https://github.com/RobTillaart/DHT20/releases)


# DHT20

Arduino library for I2C DHT20 temperature and humidity sensor.


## Description

The library must be initiated by calling the **begin()** function, 
or **begin(dataPin, clockPin)** for **ESP32** and similar platforms.

Thereafter one has to call the **read()** function to do the actual reading,
and with **getTemperature()** and **getHumidity()** to get the measured values.
Calling these latter again will return the same values until a new **read()** is called.

The **read()** call of this sensor is blocking for 80+ milliseconds (datasheet 7.4)
so the library also has a asynchronous interface. See below.

Since 0.1.3 and 0.1.4 the performance of **read()** has been optimized, 
still blocking but less long for about 45 milliseconds.


### 0.2.0

In #8 a bug is described that the sensor "freezes".
Cause is not well understood.

Two solutions / workarounds are found:
- call **resetSensor()** before EVERY **read()**. 
This is the preferred solution.
- use **Wire.setClock(200000)** 100 K and lower speeds freezes the sensor.
With clock set to 200 K and above the sensor seems to work for longer periods.
Tested several speeds on UNO, no pull ups, < 10 cm wire.

Note: setting the I2C clock possibly interferes with other devices on the bus,
so it is not a solution in the end.

The 0.2.0 version embeds the **resetSensor()** into **requestData()** to 
reset the sensor if needed in both synchronous and asynchronous calls.
This keeps it easy for the user. The reads are 1-2 ms slower than 0.1.4. 
Still far below the 80 ms mentioned in the datasheet. (< 50 ms).


### Connection

Always check datasheet!

Front view
```
          +--------------+
  VDD ----| 1            |
  SDA ----| 2    DHT20   |
  GND ----| 3            |
  SCL ----| 4            |
          +--------------+
```

### Tested

Verified to work with Arduino UNO and ESP32.
Please let me know if other platforms work (or not).


## Interface


### Constructor

- **DHT20(TwoWire \*wire = &Wire)** constructor, using a specific Wire (I2C bus).
- **bool begin(uint8_t dataPin, uint8_t clockPin)** begin for ESP32 et al, to set I2C bus pins.
- **bool begin()** initializer for non ESP32. Returns true if connected.
- **bool isConnected()** returns true if the address of the DHT20 can be seen on the I2C bus.


### Core

- **int8_t read()** read the sensor and store the values internally. 
It returns the status of the read which should be 0.
- **float getHumidity()** returns last Humidity read.
- **float getTemperature()** returns last Temperature read.


### Offset

- **void setHumOffset(float offset)** set an offset to calibrate (1st order) the sensor.
- **float getHumOffset()** return current offset, default 0.
- **void setTempOffset(float offset)** set an offset to calibrate (1st order) the sensor.
- **float getTempOffset()** return current offset, default 0.


### Asynchronous interface

There are two timings that need to be considered:
- time between requests = 1000 ms
- time between request and data ready = 80 ms

The async interface allows one to continue processing after a **requestData()** has been made. 
Note that there should be at least **1000 milliseconds** between subsequent requests.

With **bool isMeasuring()** one can check if a new measurement is ready.
If so the sensor can be read with **readData()**.

To interpret the read bits to temperature, humidity and status one needs to call **convert()** as last step.


- **int requestData()** signals the sensor to make a new measurement.
Note there must be at least 1000 milliseconds between requests!
- **int readData()** does the actual reading of the data.
- **int convert()** converts the read bits to temperature and humidity.

See the example **DHT20_async.ino**


### Status

- **uint8_t readStatus()** forced read of the status only.
This function blocks a few milliseconds to optimize communication.
- **bool isCalibrated()** idem, wrapper around **readStatus()**
- **bool isMeasuring()** idem, wrapper around **readStatus()**
- **bool isIdle()** idem, wrapper around **readStatus()**
- **int internalStatus()** returns the internal status of the sensor. (for debug).

|  status bit  |  meaning                   |
|:------------:|:---------------------------|
|    7         |  1 = measurement, 0 = idle |
|  6 - 4       |  unknown                   |
|    3         |  1 = calibrated, 0 = not   |
|  2 - 0       |  unknown                   |


#### Experimental 0.1.4 resetSensor

Use with care!

- **uint8_t resetSensor()** if at startup the sensor does not return a status of 0x18, 
three registers 0x1B, 0x1C and 0x1E need to be reset. 
See datasheet 7.4 Sensor Reading Process, point 1.
There is no documentation about the meaning of these registers.
The code is based upon example code for the AHT20 (from manufacturer).

The call is needed to get the **read()** working well so it has been embedded into
the read calls. (0.2.0)


### Timing

- **uint32_t lastRead()** last time the sensor is read in milliseconds since start.
- **uint32_t lastRequest()** last time a request is made to make a measurement.


### Return codes

| name                        |  value  |  notes  |
|:----------------------------|:-------:|:--------|
| DHT20_OK                    |    00   |  OK
| DHT20_ERROR_CHECKSUM        |   -10   |  values might be OK if they are like recent previous ones.
| DHT20_ERROR_CONNECT         |   -11   |  check connection
| DHT20_MISSING_BYTES         |   -12   |  check connection
| DHT20_ERROR_BYTES_ALL_ZERO  |   -13   |  check connection
| DHT20_ERROR_READ_TIMEOUT    |   -14   |
| DHT20_ERROR_LASTREAD        |   -15   |  wait 1 second between reads


## Operation

See examples


## Future

#### must
- improve documentation.
- investigate the bug from #8 further

#### should


#### could
- improve unit tests.
- investigate 
  - sensor calibration (website aosong?)
- investigate optimizing timing in readStatus()
  - delay(1) ==> microSeconds(???).
- connected flag?

#### won't

- **void setIgnoreChecksum(bool = false)** ignore checksum flag speeds up communication a bit
- **bool getIgnoreChecksum()** get checksum flag. for completeness.
- 




CHANGELOG.md
# Change Log DHT20

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## [0.2.0] - 2022-10-30
- add changelog.md
- add rp2040 to build-CI
- add workaround for #8 to readme.md
- fix requestFrom() ambiguity
- fix #8 - embed **resetSensor()** into **read()**.
  This makes reads slightly slower.

----

## [0.1.4] - 2022-09-18
- add resetSensor() code.
- add comments in .h file
- add examples
- stabilize readStatus()
- update readme.md

## [0.1.3] - 2022-09-17
- add wrapper status functions
- improve performance read()
- refactor, update readme.md

## [0.1.2] - 2022-09-16
- fix #4 DHT20_ERROR_BYTES_ALL_ZERO error condition.
- fix keywords
- add readStatus()  fix _readStatus()
- add setWireTimeout(250000, true);  //  in comments

## [0.1.1] - 2022-09-10
- add hardware schema to readme.md.
- fix async interface (first version)

## [0.1.0] - 2022-01-11
- initial version (based upon DHT20 datasheet)



test/unit_test_001.cpp
//
//    FILE: unit_test_001.cpp
//  AUTHOR: Rob Tillaart
//    DATE: 2022-01-11
// PURPOSE: unit tests for the DHT20 library
//          https://github.com/RobTillaart/DHT20
//          https://github.com/Arduino-CI/arduino_ci/blob/master/REFERENCE.md
//

// supported assertions
// https://github.com/Arduino-CI/arduino_ci/blob/master/cpp/unittest/Assertion.h#L33-L42
// ----------------------------
// assertEqual(expected, actual)
// assertNotEqual(expected, actual)
// assertLess(expected, actual)
// assertMore(expected, actual)
// assertLessOrEqual(expected, actual)
// assertMoreOrEqual(expected, actual)
// assertTrue(actual)
// assertFalse(actual)
// assertNull(actual)
// assertNotNull(actual)


#include <ArduinoUnitTests.h>


#include "Wire.h"
#include "DHT20.h"


unittest_setup()
{
  fprintf(stderr, "DHT20_LIB_VERSION: %s \n", (char *) DHT20_LIB_VERSION);
}

unittest_teardown()
{
}


unittest(test_constants)
{
  assertEqual(  0, DHT20_OK);
  assertEqual(-10, DHT20_ERROR_CHECKSUM);
  assertEqual(-11, DHT20_ERROR_CONNECT);
  assertEqual(-12, DHT20_MISSING_BYTES);
  assertEqual(-13, DHT20_ERROR_BYTES_ALL_ZERO);
  assertEqual(-14, DHT20_ERROR_READ_TIMEOUT);
  assertEqual(-15, DHT20_ERROR_LASTREAD);
}


unittest(test_constructor)
{
  DHT20 DHT(&Wire);

  assertEqualFloat(0, DHT.getTemperature(), 0.001);
  assertEqualFloat(0, DHT.getHumidity(), 0.001);
  assertEqualFloat(0, DHT.getTempOffset(), 0.001);
  assertEqualFloat(0, DHT.getHumOffset(), 0.001);

  DHT.begin();
  assertEqual(DHT20_ERROR_CONNECT, DHT.read());

  //  assertEqualFloat(0, DHT.getTemperature(), 0.001);
  //  assertEqualFloat(0, DHT.getHumidity(), 0.001);
  assertEqualFloat(0, DHT.getTempOffset(), 0.001);
  assertEqualFloat(0, DHT.getHumOffset(), 0.001);
}


unittest(test_offset)
{
  DHT20 DHT(&Wire);
  DHT.begin();

  assertEqualFloat(0, DHT.getTempOffset(), 0.001);
  assertEqualFloat(0, DHT.getHumOffset(), 0.001);

  for (float offset = -1.5; offset < 1.5; offset += 0.5)
  {
    DHT.setHumOffset(offset);
    DHT.setTempOffset(offset);

    assertEqualFloat(offset, DHT.getTempOffset(), 0.001);
    assertEqualFloat(offset, DHT.getHumOffset(), 0.001);

    assertEqualFloat(offset, DHT.getTemperature(), 0.001);
    assertEqualFloat(offset, DHT.getHumidity(), 0.001);
  }
}


unittest_main()

// --------


DHT20.h
#pragma once
//
//    FILE: DHT20.h
//  AUTHOR: Rob Tillaart
// PURPOSE: Arduino library for DHT20 I2C temperature and humidity sensor.
// VERSION: 0.1.4
// HISTORY: See DHT20.cpp
//     URL: https://github.com/RobTillaart/DHT20
//

//  Always check datasheet - front view
//
//          +--------------+
//  VDD ----| 1            |
//  SDA ----| 2    DHT20   |
//  GND ----| 3            |
//  SCL ----| 4            |
//          +--------------+


#include "Arduino.h"
#include "Wire.h"

#define DHT20_LIB_VERSION                    (F("0.1.4"))

#define DHT20_OK                             0
#define DHT20_ERROR_CHECKSUM                -10
#define DHT20_ERROR_CONNECT                 -11
#define DHT20_MISSING_BYTES                 -12
#define DHT20_ERROR_BYTES_ALL_ZERO          -13
#define DHT20_ERROR_READ_TIMEOUT            -14
#define DHT20_ERROR_LASTREAD                -15


class DHT20
{
public:
  //  CONSTRUCTOR
  //  fixed address 0x38
  DHT20(TwoWire *wire = &Wire);

  //  start the I2C
#if defined(ESP8266) || defined(ESP32)
  bool     begin(const uint8_t dataPin, const uint8_t clockPin);
#endif
  bool     begin();
  bool     isConnected();


  //  ASYNCHRONUOUS CALL
  //  trigger acquisition.
  int      requestData();
  //  read the raw data.
  int      readData();
  //  converts raw databits to temperature and humidity.
  int      convert();


  //  SYNCHRONUOUS CALL
  //  blocking read call to read + convert data
  int      read();
  //  access the converted temperature & humidity
  float    getHumidity();
  float    getTemperature();


  //  OFFSET  1st order adjustments
  void     setHumOffset(float offset);
  void     setTempOffset(float offset);
  float    getHumOffset();
  float    getTempOffset();


  //  READ STATUS
  uint8_t  readStatus();
  //  3 wrapper functions around readStatus()
  bool     isCalibrated();
  bool     isMeasuring();
  bool     isIdle();
  //  status from last read()
  int      internalStatus();


  //  TIMING
  uint32_t lastRead();
  uint32_t lastRequest();


  //  RESET  (new since 0.1.4)
  //  use with care 
  //  returns number of registers reset => must be 3
  //  3     = OK
  //  0,1,2 = error.
  //  255   = no reset needed.
  //  See datasheet 7.4 Sensor Reading Process, point 1
  //  use with care 
  uint8_t  resetSensor();


private:
  float    _humidity;
  float    _temperature;
  float    _humOffset;
  float    _tempOffset;

  uint8_t  _status;
  uint32_t _lastRequest;
  uint32_t _lastRead;
  uint8_t  _bits[7];

  uint8_t  _crc8(uint8_t *ptr, uint8_t len);

  //  use with care
  void     _resetRegister(uint8_t reg);

  TwoWire* _wire;
};


// -- END OF FILE --



DHT20.cpp
//
//    FILE: DHT20.cpp
//  AUTHOR: Rob Tillaart
// VERSION: 0.1.4
// PURPOSE: Arduino library for DHT20 I2C temperature and humidity sensor.
//
// HISTORY:
//  0.1.0   2022-01-11  initial version (based upon DHT20 datasheet)
//  0.1.1   2022-09-10  add hardware schema to readme.md.
//                      fix async interface (first version)
//  0.1.2   2022-09-16  fix #4 DHT20_ERROR_BYTES_ALL_ZERO error condition.
//                      fix keywords
//                      add readStatus()  fix _readStatus()
//                      add setWireTimeout(250000, true);  // in comments
//  0.1.3   2022-09-17  add wrapper status functions
//                      improve performance read()
//                      refactor, update readme.md
//  0.1.4   2022-09-18  add initialization code.
//                      add comments in .h file
//                      add examples
//                      stabilize readStatus()
//                      update readme.md


#include "DHT20.h"


//  set DHT20_WIRE_TIME_OUT to 0 to disable.
//  note this timeout is commented in code below.
#define DHT20_WIRE_TIME_OUT         250000    //  microseconds

const uint8_t DHT20_ADDRESS = 0x38;


DHT20::DHT20(TwoWire *wire)
{
  _wire        = wire;
  //  reset() ?
  _temperature = 0;
  _humidity    = 0;
  _humOffset   = 0;
  _tempOffset  = 0;
  _status      = 0;
  _lastRequest = 0;
  _lastRead    = 0;
}


bool DHT20::begin()
{
  _wire->begin();
  //  _wire->setWireTimeout(DHT20_WIRE_TIME_OUT, true);
  return isConnected();
}


#if defined(ESP8266) || defined(ESP32)
bool DHT20::begin(const uint8_t dataPin, const uint8_t clockPin)
{
  if ((dataPin < 255) && (clockPin < 255))
  {
    _wire->begin(dataPin, clockPin);
  } else {
    _wire->begin();
  }
  //  _wire->setWireTimeout(DHT20_WIRE_TIME_OUT, true);
  return isConnected();
}
#endif


bool DHT20::isConnected()
{
  _wire->beginTransmission(DHT20_ADDRESS);
  int rv = _wire->endTransmission();
  return rv == 0;
}


//  See datasheet 7.4 Sensor Reading Process, point 1
//  use with care.
uint8_t DHT20::resetSensor()
{
  uint8_t count = 255;
  if ((readStatus() & 0x18) != 0x18)
  {
    count++;
    if (_resetRegister(0x1B)) count++;
    if (_resetRegister(0x1C)) count++;
    if (_resetRegister(0x1E)) count++;
    delay(10);
  }
  return count;
}


////////////////////////////////////////////////
//
//  READ THE SENSOR
//
int DHT20::read()
{
  //  do not read to fast
  if (millis() - _lastRead < 1000)
  {
    return DHT20_ERROR_LASTREAD;
  }

  int status = requestData();
  if (status < 0) return status;
  //  wait for measurement ready
  while (isMeasuring())
  {
    yield();
  }
  //  read the measurement
  status = readData();
  if (status < 0) return status;

  //  convert it to meaningfull data
  return convert();
}


int DHT20::requestData()
{
  //  GET CONNECTION
  _wire->beginTransmission(DHT20_ADDRESS);
  _wire->write(0xAC);
  _wire->write(0x33);
  _wire->write(0x00);
  int rv = _wire->endTransmission();

  _lastRequest = millis();
  return rv;
}


int DHT20::readData()
{
  //  GET DATA
  const uint8_t length = 7;
  int bytes = _wire->requestFrom(DHT20_ADDRESS, length);

  if (bytes == 0)     return DHT20_ERROR_CONNECT;
  if (bytes < length) return DHT20_MISSING_BYTES;

  bool allZero = true;
  for (int i = 0; i < bytes; i++)
  {
    _bits[i] = _wire->read();
    //  if (_bits[i] < 0x10) Serial.print(0);
    //  Serial.print(_bits[i], HEX);
    //  Serial.print(" ");
    allZero = allZero && (_bits[i] == 0);
  }
  //  Serial.println();
  if (allZero) return DHT20_ERROR_BYTES_ALL_ZERO;

  _lastRead = millis();
  return bytes;
}


int DHT20::convert()
{
  //  CONVERT AND STORE
  _status      = _bits[0];
  uint32_t raw = _bits[1];
  raw <<= 8;
  raw += _bits[2];
  raw <<= 4;
  raw += (_bits[3] >> 4);
  _humidity = raw * 9.5367431640625e-5;   // ==> / 1048576.0 * 100%;

  raw = (_bits[3] & 0x0F);
  raw <<= 8;
  raw += _bits[4];
  raw <<= 8;
  raw += _bits[5];
  _temperature = raw * 1.9073486328125e-4 - 50;  //  ==> / 1048576.0 * 200 - 50;

  //  TEST CHECKSUM
  uint8_t _crc = _crc8(_bits, 6);
  //  Serial.print(_crc, HEX);
  //  Serial.print("\t");
  //  Serial.println(_bits[6], HEX);
  if (_crc != _bits[6]) return DHT20_ERROR_CHECKSUM;

  return DHT20_OK;
}


////////////////////////////////////////////////
//
//  TEMPERATURE & HUMIDITY & OFFSET
//
float DHT20::getHumidity()
{
  return _humidity + _humOffset;
};


float DHT20::getTemperature()
{
  return _temperature + _tempOffset;
};


void DHT20::setHumOffset(float offset)
{
  _humOffset  = offset;
};


void DHT20::setTempOffset(float offset)
{
  _tempOffset = offset;
};


float DHT20::getHumOffset()
{
  return _humOffset;
};


float DHT20::getTempOffset()
{
  return _tempOffset;
};


////////////////////////////////////////////////
//
//  STATUS
//
uint8_t DHT20::readStatus()
{
  _wire->beginTransmission(DHT20_ADDRESS);
  _wire->write(0x71);
  _wire->endTransmission();
  delay(1);  //  needed to stabilize timing
  _wire->requestFrom(DHT20_ADDRESS, (uint8_t)1);
  delay(1);  //  needed to stabilize timing
  return (uint8_t) _wire->read();
}


bool DHT20::isCalibrated()
{
  return (readStatus() & 0x08) == 0x08;
}


bool DHT20::isMeasuring()
{
  return (readStatus() & 0x80) == 0x80;
}


bool DHT20::isIdle()
{
  return (readStatus() & 0x80) == 0x00;
}


int DHT20::internalStatus()
{
  return _status;
};


////////////////////////////////////////////////
//
//  TIMING
//
uint32_t DHT20::lastRead()
{
  return _lastRead;
};


uint32_t DHT20::lastRequest()
{
  return _lastRequest;
};


////////////////////////////////////////////////
//
//  PRIVATE
//
uint8_t DHT20::_crc8(uint8_t *ptr, uint8_t len)
{
  uint8_t crc = 0xFF;
  while(len--)
  {
    crc ^= *ptr++;
    for (uint8_t i = 0; i < 8; i++)
    {
      if (crc & 0x80)
      {
        crc <<= 1;
        crc ^= 0x31;
      }
      else
      {
        crc <<= 1;
      }
    }
  }
  return crc;
}


//  Code based on demo code sent by www.aosong.com
//  no further documentation.
//  0x1B returned 18, 0, 4
//  0x1C returned 18, 65, 0
//  0x1E returned 18, 8, 0
//    18 seems to be status register
//    other values unknown.
bool DHT20::_resetRegister(uint8_t reg)
{
  uint8_t value[3];
  _wire->beginTransmission(DHT20_ADDRESS);
  _wire->write(reg);
  _wire->write(0x00);
  _wire->write(0x00);
  if (_wire->endTransmission() != 0) return false;
  delay(5);

  int bytes = _wire->requestFrom(DHT20_ADDRESS, 3); 
  for (int i = 0; i < bytes; i++)
  {
    value[i] = _wire->read();
    //  Serial.println(value[i], HEX);
  }
  delay(10);

  _wire->beginTransmission(DHT20_ADDRESS);
  _wire->write(0xB0 | reg);
  _wire->write(value[1]);
  _wire->write(value[2]);
  if (_wire->endTransmission() != 0) return false;
  delay(5);
  return true;
}


// -- END OF FILE --

